### PR TITLE
[Scrolller] Add Extractor

### PR DESCRIPTION
**Pull request by LunarFang416**
### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

---

### Description of your *pull request* and other information

Extractor added for scrolller.com.

This PR addresses the new extractor request mentioned in #3635 . The new extractor is now able to download videos from scrolller.com which is a video/image viewing platform.

The code passes all tests and heeds to flake8 formatting and yt-dlp coding conventions.

(ignore the other 2 random commits, those were done on accident xd)

### CODE REVIEW

**Comment by pukkandan** [comment]
Read contributing guidelines. The whole extractor is wrong

**Comment by LunarFang416** [comment]
@pukkandan I apologize for my mistakes and my ignorance.

I have a few questions. On the website there are videos and images, how should i proceed with those? Should I add the image link as simply the url?

Also, could you give a little bit of guidance on what should be returned from the extractor, after reading the guidelines I am a little bit confused.

Thanks in advance, and sorry for bothering you.

**Comment by pukkandan** [comment]
> On the website there are videos and images, how should i proceed with those? Should I add the image link as simply the url?

Don't support images

> Also, could you give a little bit of guidance on what should be returned from the extractor, after reading the guidelines I am a little bit confused.

What exactly are you confused about?

**Comment by LunarFang416** [comment]
@pukkandan 
1. The only way to tell if there is an image is through the image source url (as there is a .jpg in the urk) which is obtained after sending a request to the graphql server. If that is the case what error should i return.
2. After querying the graphql api, this is the following results that i get:
```json
{
      "__typename":"SubredditPost",
      "id":7793294,
      "url":"/its-all-fun-and-games-cco8jjmoh7",
      "title":"It's all fun and games...",
      "subredditTitle":"CatsISUOTTATFO",
      "subredditUrl":"/r/CatsISUOTTATFO",
      "redditPath":"/r/CatsISUOTTATFO/comments/aspuho/its_all_fun_and_games/",
      "isNsfw":false,
      "albumUrl":null,
      "hasAudio":false,
      "fullLengthSource":null,
      "gfycatSource":null,
      "redgifsSource":null,
      "mediaSources":[
         {
            "url":"https://atto.scrolller.com/its-all-fun-and-games-3amk9vg7m3-540x649.jpg",
            "width":540,
            "height":649,
            "isOptimized":true
         },
         {
            "url":"https://yocto.scrolller.com/its-all-fun-and-games-3amk9vg7m3.mp4",
            "width":960,
            "height":1154,
            "isOptimized":false
         }
      ]
   },
```
From here what is confusing me is the source of the video. I can tell from the json object that the url for the video is `https://yocto.scrolller.com/its-all-fun-and-games-3amk9vg7m3.mp4`, so do I set it to a `url` field in the dictionary and return it? Or do i need to add it to the formats field? The formats filed requires and extensions, so would i have to parse the url to get the extension? 

I apologize if I have the wrong idea about things.

**Comment by pukkandan** [comment]
> The only way to tell if there is an image is through the image source url (as there is a .jpg in the urk) which is obtained after sending a request to the graphql server. If that is the case what error should i return.

In the example you have given, there is an image and a video. In this case, you can simply return the image as `thumbnails` and the video as `formats`. For URLs without videos, you can either `raise ExtractorError('There is no video', expected=True)` or simply return without a `formats`, whichever is convinient (In the latter case, core code will take care of error handling)


> so do I set it to a url field in the dictionary and return it? Or do i need to add it to the formats field?

Do all videos only have one URL each? If so, you can use `url`. Otherwise, you need to use `formats`

> The formats filed requires and extensions, so would i have to parse the url to get the extension?

`url` is the only mandatory field in formats. So you don't need to parse the `ext`

File: yt_dlp/extractor/scrolller.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+            "id": str_or_none(video.get("id")),
+            "thumbnail": traverse_obj(video, "mediaSources")[0].get("url"),
+            "formats": traverse_obj(video, "mediaSources"),
+            "age_limit": 18 if video.get("isNsfw") else 0
+        })
+
+        del video['isNsfw']
+        del video['mediaSources']
+
+        return video
```
Comment (Ashish0804) [inline yt_dlp/extractor/scrolller.py:128]: Do not do this.
Explicitly set all the fields yourself.

File: yt_dlp/extractor/scrolller.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+                    redditPath
+                    isNsfw
+                    mediaSources{
+                        url
+                        width
+                        height
+                    }
+                }
+            }''' % (video_id)
+        }).encode(), headers={
```
Comment (Ashish0804) [inline yt_dlp/extractor/scrolller.py:114]: move data out.
This is hard to read

File: yt_dlp/extractor/scrolller.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+            'Content-Type': 'application/json',
+        })['data']['getSubredditPost']
+
+        video.update({
+            "id": str_or_none(video.get("id")),
+            "thumbnail": traverse_obj(video, "mediaSources")[0].get("url"),
+            "formats": traverse_obj(video, "mediaSources"),
```
Comment (Ashish0804) [inline yt_dlp/extractor/scrolller.py:121]: sort formats

File: yt_dlp/extractor/scrolller.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+                        height
+                    }
+                }
+            }''' % (video_id)
+        }).encode(), headers={
+            'Content-Type': 'application/json',
+        })['data']['getSubredditPost']
+
+        video.update({
+            "id": str_or_none(video.get("id")),
```
Comment (Ashish0804) [inline yt_dlp/extractor/scrolller.py:119]: Use id from url regex

**Review by Ashish0804 — CHANGES_REQUESTED** [review]

**Comment by LunarFang416** [comment]
@Ashish0804 Thanks you for the feedback, I appreciate it! I made the changes in the latest commit.

File: yt_dlp/extractor/scrolller.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+        query = {
+            'query':
+            '''{
+                getSubredditPost(url:"/%s"){
+                    id
+                    title
+                    subredditTitle
+                    subredditUrl
+                    redditPath
```
Comment (Ashish0804) [inline yt_dlp/extractor/scrolller.py:67]: U are not using these

try to see if this works after removing these

File: yt_dlp/extractor/scrolller.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+                    isNsfw
+                    mediaSources{
+                        url
+                        width
+                        height
```
Comment (Ashish0804) [inline yt_dlp/extractor/scrolller.py:72]: is it guaranteed these will always be int?

File: yt_dlp/extractor/scrolller.py [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+            }''' % (video_id)
+        }
+
+        video = self._download_json(
+            "https://api.scrolller.com/api/v2/graphql", video_id, data=json.dumps(query).encode(), headers={'Content-Type': 'application/json'})['data']['getSubredditPost']
+
+        thumbnails = []
+        formats = []
+
+        for source in video.get("mediaSources"):
```
Comment (Ashish0804) [inline yt_dlp/extractor/scrolller.py:84]:
```suggestion
        for source in video.get("mediaSources") or []:
```

File: yt_dlp/extractor/scrolller.py [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+                    mediaSources{
+                        url
+                        width
+                        height
+                    }
+                }
+            }''' % (video_id)
+        }
+
+        video = self._download_json(
```
Comment (Ashish0804) [inline yt_dlp/extractor/scrolller.py:78]: rename this to json_data or video_json or video_data or info_data (preferred) 

Naming it video is kinda confusing at first glance

File: yt_dlp/extractor/scrolller.py (hunk 6 above)
Comment (LunarFang416) [reply yt_dlp/extractor/scrolller.py:69]: I queried the graphQl api to get some data about the mutations and schema it uses and also get some data about the entry fields. The following fields indicate that width and height are scalar values, but I still added the int_or_none() guard to make sure there are no errors:

```json
{
      "name": "width",
      "args": [],
      "type": {
          "kind": "NON_NULL",
          "name": null,
          "ofType": {
              "kind": "SCALAR",
              "name": "Int",
              "ofType": null
          }
      },
      "isDeprecated": false,
      "deprecationReason": null
  },
  {
      "name": "height",
      "args": [],
      "type": {
          "kind": "NON_NULL",
          "name": null,
          "ofType": {
              "kind": "SCALAR",
              "name": "Int",
              "ofType": null
          }
      },
      "isDeprecated": false,
      "deprecationReason": null
  },
```

File: yt_dlp/extractor/scrolller.py [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+                    "url": source.get("url"),
+                    "width": int_or_none(source.get("width")),
+                    "height": int_or_none(source.get("height")),
+                })
+            else:
+                formats.append({
+                    "url": source.get("url"),
+                    "width": int_or_none(source.get("width")),
+                    "height": int_or_none(source.get("height")),
+                })
```
Comment (pukkandan) [inline yt_dlp/extractor/scrolller.py:93]: this is a bad idea. You might end up with random data as thumbnails. Better to hard-code the supported thumbnail extensions imo

File: yt_dlp/extractor/scrolller.py [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+                        url
+                        width
+                        height
+                    }
+                }
+            }''' % (video_id)
+        }
+
+        video_data = self._download_json(
+            "https://api.scrolller.com/api/v2/graphql", video_id, data=json.dumps(query).encode(), headers={'Content-Type': 'application/json'})['data']['getSubredditPost']
```
Comment (pukkandan) [inline yt_dlp/extractor/scrolller.py:76]: line too long

File: yt_dlp/extractor/scrolller.py [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+                    title
+                    isNsfw
+                    mediaSources{
+                        url
+                        width
+                        height
+                    }
+                }
+            }''' % (video_id)
+        }
```
Comment (pukkandan) [inline yt_dlp/extractor/scrolller.py:73]:
```suggestion
        query = {
            'query': '''{
                getSubredditPost(url:"/%s"){
                    id
                    title
                    isNsfw
                    mediaSources{
                        url
                        width
                        height
                    }
                }
            }''' % video_id
        }
```

File: yt_dlp/extractor/scrolller.py [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+
+        video_data = self._download_json(
+            "https://api.scrolller.com/api/v2/graphql", video_id, data=json.dumps(query).encode(), headers={'Content-Type': 'application/json'})['data']['getSubredditPost']
+
+        thumbnails = []
+        formats = []
+
+        for source in video_data.get("mediaSources") or []:
```
Comment (pukkandan) [inline yt_dlp/extractor/scrolller.py:81]:
```suggestion
        for source in video_data['mediaSources']:
```

**Review by pukkandan — CHANGES_REQUESTED** [review]
use single quotes everywhere

File: yt_dlp/extractor/scrolller.py (hunk 9 above)
Comment (LunarFang416) [reply yt_dlp/extractor/scrolller.py:93]: That makes sense. I've only seen jpg and png thumbnails, so I'll only let those thumbnails pass through.

File: yt_dlp/extractor/scrolller.py [hunk 13]
```diff
@@ -0,0 +1,107 @@
+from .common import InfoExtractor
+from ..utils import (determine_ext, int_or_none)
+import json
```
Comment (pukkandan) [inline yt_dlp/extractor/scrolller.py:3]:
```suggestion
import json

from .common import InfoExtractor
from ..utils import determine_ext, int_or_none
```

File: yt_dlp/extractor/scrolller.py [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+
+        video_data = self._download_json(
+            'https://api.scrolller.com/api/v2/graphql', video_id, data=json.dumps(query).encode(),
+            headers={'Content-Type': 'application/json'})['data']['getSubredditPost']
+
+        thumbnails = []
+        formats = []
+
+        for source in video_data['mediaSources']:
+            if determine_ext(source.get('url')) not in ('.jpg', '.png'):
```
Comment (pukkandan) [inline yt_dlp/extractor/scrolller.py:83]: `not in`? did you test this code?

File: yt_dlp/extractor/scrolller.py [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+                thumbnails.append({
+                    'url': source.get('url'),
```
Comment (pukkandan) [inline yt_dlp/extractor/scrolller.py:85]:
```suggestion
                    'url': source['url'],
```

File: yt_dlp/extractor/scrolller.py [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+                    'width': int_or_none(source.get('width')),
+                    'height': int_or_none(source.get('height')),
+                })
+            else:
+                formats.append({
+                    'url': source.get('url'),
```
Comment (pukkandan) [inline yt_dlp/extractor/scrolller.py:91]:
```suggestion
            elif source.get('url'):
                formats.append({
                    'url': source['url'],
```

**Review by pukkandan — CHANGES_REQUESTED** [review]
you made a mistake in the merge. Add the extractor import back in `_extractors.py` to fix it

**Comment by LunarFang416** [comment]
@pukkandan I made the fixes. I kept messing stuff up, so i'm very sorry for the inconveniences I may have caused you through this PR.

File: yt_dlp/extractor/scrolller.py [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+                    id
+                    title
+                    isNsfw
+                    mediaSources{
+                        url
+                        width
+                        height
+                    }
+                }
+            }''' % (video_id)
```
Comment (pukkandan) [inline yt_dlp/extractor/scrolller.py:73]:
```suggestion
            'query': '''{
                getSubredditPost(url:"/%s"){
                    id
                    title
                    isNsfw
                    mediaSources{
                        url
                        width
                        height
                    }
                }
            }''' % video_id
```

File: yt_dlp/extractor/scrolller.py [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+        }
+
+        video_data = self._download_json(
+            'https://api.scrolller.com/api/v2/graphql', video_id, data=json.dumps(query).encode(),
+            headers={'Content-Type': 'application/json'})['data']['getSubredditPost']
+
+        thumbnails = []
+        formats = []
+
```
Comment (pukkandan) [inline yt_dlp/extractor/scrolller.py:82]:
```suggestion
        formats, thumbnails = [], []
```

File: yt_dlp/extractor/scrolller.py [hunk 19]
```diff
@@ -0,0 +1,108 @@
...
+        for source in video_data['mediaSources']:
+            if determine_ext(source['url']) in ('jpg', 'png'):
```
Comment (pukkandan) [inline yt_dlp/extractor/scrolller.py:84]:
```suggestion
            if determine_ext(source.get('url')) in ('jpg', 'png'):
```

File: yt_dlp/extractor/scrolller.py [hunk 20]
```diff
@@ -0,0 +1,108 @@
+import json
+
+from .common import InfoExtractor
+from ..utils import (determine_ext, int_or_none)
```
Comment (pukkandan) [inline yt_dlp/extractor/scrolller.py:4]:
```suggestion
from ..utils import determine_ext, int_or_none
```

**Review by pukkandan — CHANGES_REQUESTED** [review]

**Pull request merged**